src/performance/index.ts

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2021 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { PerformanceEntryNames } from "./entry-names";
+
+const START_PREFIX = "start:";
+const STOP_PREFIX = "stop:";
+
+export {
+    PerformanceEntryNames,
+}
+
+interface GetEntriesOptions {
+    name?: string,
+    type?: string,
+}
+
+type PerformanceCallbackFunction = (entry: PerformanceEntry[]) => void;
+
+interface PerformanceDataListener {
+    entryNames?: string[],
+    callback: PerformanceCallbackFunction
+}
+
+let listeners: PerformanceDataListener[] = [];
+const entries: PerformanceEntry[] = [];
+
+export default class PerformanceMonitor {
+    /**
+     * Starts a performance recording
+     * @param name Name of the recording
+     * @param id Specify an identifier appended to the measurement name
+     * @returns {void}
+     */
+    static start(name: string, id?: string): void {
```

Comment:
In a multi-account world, `static` gets hard to manage. If we're not concerned about linking performance to account size metrics, then this is fine, but I'd probably still recommend a `.instance` getter as this is our stepping stone to Dependency Injection or per-client accessors.